File: ios/Packages/FatwaBotKit/Sources/NetworkingKit/APIClient.swift
import CoreKit
import Foundation

/// Typed endpoint description; the full catalog lives in Endpoints.swift.
public struct Endpoint<Response: Decodable & Sendable>: Sendable {
    public let path: String
    public let query: [URLQueryItem]

    public init(path: String, query: [URLQueryItem] = []) {
        self.path = path
        self.query = query
    }
}

public struct ClientContext: Sendable {
    public let platform: String
    public let appVersion: String
    public let locale: String

    public init(platform: String = "ios", appVersion: String, locale: String) {
        self.platform = platform
        self.appVersion = appVersion
        self.locale = locale
    }
}

public enum APIError: Error, Equatable {
    case invalidURL
    case transport(String)
    case server(statusCode: Int, code: String?)
    case decoding(String)
}

/// Read path of the versioned REST API (ADR-0002). GET-only in M0; authenticated
/// writes arrive with /v1/auth in M1.
public protocol APIClientProtocol: Sendable {
    func get<Response: Decodable & Sendable>(_ endpoint: Endpoint<Response>) async throws -> Response
}

public final class APIClient: APIClientProtocol {
    private let baseURL: URL
    private let context: ClientContext
    private let session: URLSession

    public init(baseURL: URL, context: ClientContext, session: URLSession = .shared) {
        self.baseURL = baseURL
        self.context = context
        self.session = session
    }

    public func get<Response: Decodable & Sendable>(_ endpoint: Endpoint<Response>) async throws -> Response {
        guard var components = URLComponents(
            url: baseURL.appendingPathComponent(endpoint.path),
            resolvingAgainstBaseURL: false
        ) else { throw APIError.invalidURL }
        if !endpoint.query.isEmpty { components.queryItems = endpoint.query }
        guard let url = components.url else { throw APIError.invalidURL }

        var request = URLRequest(url: url)
        request.setValue(context.platform, forHTTPHeaderField: "x-client-platform")
        request.setValue(context.appVersion, forHTTPHeaderField: "x-client-version")
        request.setValue(context.locale, forHTTPHeaderField: "x-client-locale")

        let data: Data
        let response: URLResponse
        do {
            (data, response) = try await session.data(for: request)
        } catch {
            throw APIError.transport(error.localizedDescription)
        }

        guard let http = response as? HTTPURLResponse else {
            throw APIError.transport("Non-HTTP response")
        }
        guard (200..<300).contains(http.statusCode) else {
            let envelope = try? JSONDecoder().decode(ErrorEnvelope.self, from: data)
            throw APIError.server(statusCode: http.statusCode, code: envelope?.error.code)
        }
        do {
            return try JSONDecoder().decode(Response.self, from: data)
        } catch {
            throw APIError.decoding(String(describing: error))
        }
    }
}

struct ErrorEnvelope: Decodable {
    struct Payload: Decodable {
        let code: String
        let message: String
    }

    let error: Payload
}

File: android/core/network/src/main/kotlin/com/fatwabot/core/network/ApiClient.kt
package com.fatwabot.core.network

import java.io.IOException
import kotlinx.coroutines.Dispatchers
import kotlinx.coroutines.withContext
import kotlinx.serialization.json.Json
import okhttp3.HttpUrl.Companion.toHttpUrl
import okhttp3.OkHttpClient
import okhttp3.Request

/** Client context headers sent on every request (parity with iOS ClientContext). */
data class ClientContext(
    val platform: String = "android",
    val appVersion: String,
    val locale: String,
)

sealed class ApiException(message: String) : Exception(message) {
    data class Transport(val detail: String) : ApiException("transport: $detail")
    data class Server(val statusCode: Int, val code: String?) : ApiException("server $statusCode ($code)")
    data class Decoding(val detail: String) : ApiException("decoding: $detail")
}

/** Pulls `error.code` out of the standard `{error:{code,message}}` envelope
 * (see backend/functions/api/http.ts apiError) — shared by [ApiClient] and
 * [AuthenticatedApiClient] so a caller can distinguish e.g. `ai_unavailable`
 * from any other 503/500, mirroring iOS `APIError.server(statusCode:code:)`. */
internal fun extractErrorCode(body: String): String? =
    runCatching { Json.parseToJsonElement(body) }.getOrNull()?.let { el ->
        runCatching {
            (el as? kotlinx.serialization.json.JsonObject)
                ?.get("error")?.let { err ->
                    (err as? kotlinx.serialization.json.JsonObject)?.get("code")
                        ?.let { c -> (c as? kotlinx.serialization.json.JsonPrimitive)?.content }
                }
        }.getOrNull()
    }

/** Read path of the versioned REST API (ADR-0002) — mirror of iOS APIClient. */
interface ApiClientProtocol {
    suspend fun getRaw(path: String, query: Map<String, String> = emptyMap()): String
}

class ApiClient(
    private val baseUrl: String,
    private val context: ClientContext,
    private val http: OkHttpClient = OkHttpClient(),
) : ApiClientProtocol {

    override suspend fun getRaw(path: String, query: Map<String, String>): String =
        withContext(Dispatchers.IO) {
            val url = baseUrl.trimEnd('/').toHttpUrl().newBuilder()
                .apply {
                    path.trim('/').split('/').forEach(::addPathSegment)
                    query.forEach { (k, v) -> addQueryParameter(k, v) }
                }
                .build()

            val request = Request.Builder()
                .url(url)
                .header("x-client-platform", context.platform)
                .header("x-client-version", context.appVersion)
                .header("x-client-locale", context.locale)
                .build()

            val response = try {
                http.newCall(request).execute()
            } catch (e: IOException) {
                throw ApiException.Transport(e.message ?: "io error")
            }
            response.use {
                val body = it.body?.string().orEmpty()
                if (!it.isSuccessful) throw ApiException.Server(it.code, extractErrorCode(body))
                body
            }
        }
}
